torchdata/datapipes/iter/util/webdataset.py

```diff
@@ -0,0 +1,66 @@
+import re
+from typing import Dict, Iterator, List, Union
+
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+
+
+def pathsplit(p):
+    """Split a path into the basename and the extensions."""
+    if "." not in p:
+        return p, ""
+    prefix, suffix = re.search(r"^(.*?)(\.[^/]*)$", p).groups()
+    return prefix, suffix
+
+
+@functional_datapipe("webdataset")
+class WebDatasetIterDataPipe(IterDataPipe[Dict]):
+    r"""
+    Iterable DataPipe that accepts stream of (path, data) tuples and aggregates them into
+    a single dictionary based on the basenames of the paths (WebDataset file convention).
+
+    File names that do not have an extension are ignored.
+
+    Args:
+        source_datapipe: a DataPipe yielding a stream of (path, data) pairs
+    Returns:
+        a DataPipe yielding a stream of dictionaries
+    Examples:
+        >>> from torchdata.datapipes.iter import FileLister, FileOpener
+        >>>
+        >>> def decode(item):
+        >>>     key, value = item
+        >>>     if key.endswith(".txt"):
+        >>>         return key, value.read().decode("utf-8")
+        >>>     if key.endswith(".bin"):
```

Comment:
It seems like the PR isn't only limited to text data though right? Would it make sense to have a more generic `decode` function available outside of the test to deal with images

Reply:
I will submit another PR for a class that helps with decoding soon. However, such a class is just there for convenience, since decoding with `map` is easy enough; you don't need it to use WebDataset.

Since it isn't essential and since a decoding class involves discussions about how to handle external dependencies and extensibility, it probably will require more discussions, that's why I think it's best to keep it to a separate PR.